### PR TITLE
[BEAM-8645] A test case for TimestampCombiner.

### PR DESCRIPTION
**Pull request by HuangLED**
Adds test case for TimestampCombiner. 

Skip tag is added (until BEAM-8657 and BEAM-8645 are resolved)

Related mailing list dicussions:  https://lists.apache.org/thread.html/d3af1f2f84a2e59a747196039eae77812b78a991f0f293c717e5f4e1@%3Cdev.beam.apache.org%3E

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ X] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by HuangLED** [comment]
R: @robertwb

**Comment by HuangLED** [comment]
R: @robertwb 

PreCommit failed once  but I don't think it is caused by my PR.  (after all we only added two tests, and they are disabled. Tested these two new cases on local machine, they worked).

**Comment by HuangLED** [comment]
Run Portable_Python PreCommit

**Comment by HuangLED** [comment]
Hi, Robert,  

     I synced with your fixing PR for combiner lifting, now LATEST works (I removed skip tag), but EARLIEST still not correct.  

     Maybe let us merge this test case, and then you can test against your up-coming PR easier. WDYT?

File: sdks/python/apache_beam/transforms/combiners_test.py [hunk 1]
```diff
@@ -480,6 +487,82 @@ def test_with_input_types_decorator_violation(self):
...
+    p = TestPipeline(options=options)
+
+    main_stream = (p
+                   | 'main TestStream' >> TestStream()
+                   .add_elements([window.TimestampedValue(('k', 100), 0)])
+                   .add_elements([window.TimestampedValue(('k', 400), 9)])
+                   .advance_watermark_to_infinity()
+                   | 'main windowInto' >> beam.WindowInto(
+                       window.FixedWindows(10),
+                       timestamp_combiner=TimestampCombiner.OUTPUT_AT_EARLIEST)
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/combiners_test.py]: Shouldn't this be LATEST?

File: sdks/python/apache_beam/transforms/combiners_test.py [hunk 2]
```diff
@@ -480,6 +487,82 @@ def test_with_input_types_decorator_violation(self):
...
+
+  class RecordFn(beam.DoFn):
+    def process(self, elm=beam.DoFn.ElementParam, ts=beam.DoFn.TimestampParam):
+      yield (elm, ts)
+
+  @unittest.skip('BEAM-8657')
+  def test_combiner_earliest(self):
+    """Test TimestampCombiner with EARLIEST."""
+    options = PipelineOptions()
+    options.view_as(StandardOptions).streaming = True
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/combiners_test.py]: Here you can just write 

`options = PipelineOptions(streaming=True)`

Similarly below.

File: sdks/python/apache_beam/transforms/combiners_test.py [hunk 3]
```diff
@@ -480,6 +487,82 @@ def test_with_input_types_decorator_violation(self):
...
+    p = TestPipeline(options=options)
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/combiners_test.py]: Prefer 

```
with TestPipeline(options=options):
    [pipeline construction
```

over using `p.run()`.

File: sdks/python/apache_beam/transforms/combiners_test.py [hunk 4]
```diff
@@ -480,6 +487,82 @@ def test_with_input_types_decorator_violation(self):
...
+
+    main_stream = (p
+                   | 'main TestStream' >> TestStream()
+                   .add_elements([window.TimestampedValue(('k', 100), 0)])
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/combiners_test.py]: I'd probably be a stronger test if the timestamps weren't the endpoints, e.g. 2 and 7. (Similarly below.)

File: sdks/python/apache_beam/transforms/combiners_test.py (hunk 1 above)
Comment (HuangLED) [reply sdks/python/apache_beam/transforms/combiners_test.py]: oops.  result from OUTPUT_AT_EARLIEST is indeed wrong, which leads to the false positive here. 

Fixed.

File: sdks/python/apache_beam/transforms/combiners_test.py (hunk 4 above)
Comment (HuangLED) [reply sdks/python/apache_beam/transforms/combiners_test.py]: Updated to be 2 and 7 respectively.

**Comment by HuangLED** [comment]
Pulled from master, and retried.  The EARLIEST case still failed, with error: 

"apache_beam.runners.direct.executor: WARNING: A task failed with exception: Failed assert: [(('k', 500), Timestamp(7))] not in [(('k', 500), Timestamp(2))] [while running 'assert per window/Match']"

**Comment by HuangLED** [comment]
would you please point out which PR in the master *Might* resolve the issue?  I can follow and trace that part a bit using this test case.

File: sdks/python/apache_beam/transforms/combiners_test.py [hunk 5]
```diff
@@ -480,6 +486,76 @@ def test_with_input_types_decorator_violation(self):
...
+                | 'main TestStream' >> TestStream()
+                .add_elements([window.TimestampedValue(('k', 100), 2)])
+                .add_elements([window.TimestampedValue(('k', 400), 7)])
+                .advance_watermark_to_infinity()
+                | 'main windowInto' >> beam.WindowInto(
+                    window.FixedWindows(10),
+                    timestamp_combiner=TimestampCombiner.OUTPUT_AT_EARLIEST)
+                | 'Combine' >> beam.CombinePerKey(sum))
+
+      records = (result | beam.ParDo(self.RecordFn()))
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/combiners_test.py]: As before, a full DoFn is not needed here, just write `beam.Map(lambda e, ts=beam.DoFn.TimestampParam: (e, ts))`

Comment (robertwb) [reply sdks/python/apache_beam/transforms/combiners_test.py]: Also, no need for the surrounding ()'s.

File: sdks/python/apache_beam/transforms/combiners_test.py [hunk 6]
```diff
@@ -480,6 +486,76 @@ def test_with_input_types_decorator_violation(self):
...
+    def process(self, elm=beam.DoFn.ElementParam, ts=beam.DoFn.TimestampParam):
+      yield (elm, ts)
+
+  @unittest.skip('BEAM-8657')
+  def test_combiner_earliest(self):
+    """Test TimestampCombiner with EARLIEST."""
+    options = PipelineOptions(streaming=True)
+    with TestPipeline(options=options) as p:
+      result = (p
+                | 'main TestStream' >> TestStream()
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/combiners_test.py]: Omit the `label >>` unless they are needed for uniqueness or clarity. These labels don't add anything to what the default labels provide.

File: sdks/python/apache_beam/transforms/combiners_test.py (hunk 5 above)
Comment (HuangLED) [reply sdks/python/apache_beam/transforms/combiners_test.py]: Ah. Thanks for the tips. :-D

File: sdks/python/apache_beam/transforms/combiners_test.py (hunk 6 above)
Comment (HuangLED) [reply sdks/python/apache_beam/transforms/combiners_test.py]: Thanks!  Will double check the style on all the other pending PRs as well.

**Comment by HuangLED** [comment]
Updated.

File: sdks/python/apache_beam/transforms/combiners_test.py [hunk 7]
```diff
@@ -480,6 +486,74 @@ def test_with_input_types_decorator_violation(self):
...
+          label='assert per window')
+
+  def test_combiner_latest(self):
+    """Test TimestampCombiner with LATEST."""
+    options = PipelineOptions(streaming=True)
+    with TestPipeline(options=options) as p:
+      result = (p
+                | TestStream()
+                .add_elements([window.TimestampedValue(('k', 100), 2)])
+                .add_elements([window.TimestampedValue(('k', 400), 7)])
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/combiners_test.py]: Turns out this happens to work because 7 is the last element.

**Comment by HuangLED** [comment]
Run Portable_Python PreCommit

**Comment by HuangLED** [comment]
Thanks for applying the unskip test. 

All test now passed. This PR is ready to be merged.

**Comment by HuangLED** [comment]
@robertwb  ping for merging this PR

**Pull request merged**